### PR TITLE
Feature/telemetry metrics

### PR DESCRIPTION
**Pull request by maple10001**
# Type of change? 

- [ ] Bug fix
- [x] New feature
- [ ] Enhancement
- [ ] Breaking change (breaks backward compatibility)

# Description, Context, Motivation 
- Add `TotalMessagesCount` in `status` API
- Add `FailedMessagesCount` in `status` API
- Add `MessagesPerSecond` in `status` API
- Add `FailedDeviceConnectionsCount` in `status` API
- Add `FailedDeviceTwinUpdatesCount` in `status` API

**Checklist:**

- [x] All tests passed
- [ ] The code follows the code style and conventions of this project
- [ ] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: Services/Concurrency/RatedCounter.cs [hunk 1]
```diff
@@ -177,6 +177,23 @@ public long GetPause()
             return false;
         }
 
+        public double GetThroughputForMessages()
+        {
+            if (this.timestamps.Count > 1)
+            {
```
Comment (avanikp) [inline Services/Concurrency/RatedCounter.cs]: This code is duplicated in LogThroughput. Suggest updating LogThroughput to call this method to calculate throughput

File: Services/Concurrency/RatedCounter.cs [hunk 2]
```diff
@@ -177,6 +177,23 @@ public long GetPause()
...
+                double speed;
+                lock (this.timestamps)
+                {
+                    long time = this.timestamps.Last() - this.timestamps.First();
+                    speed = (1000 * (double)this.timestamps.Count / time * 10) / 10;
```
Comment (avanikp) [inline Services/Concurrency/RatedCounter.cs]: MIN_TIME_UNIT

Comment (avanikp) [inline Services/Concurrency/RatedCounter.cs]: Can you make 10 as a const and add description for it

File: Services/Concurrency/RatedCounter.cs [hunk 3]
```diff
@@ -177,6 +177,23 @@ public long GetPause()
             return false;
         }
 
+        public double GetThroughputForMessages()
```
Comment (avanikp) [inline Services/Concurrency/RatedCounter.cs]: please add summary comments for the functions

        /// <summary>
        /// 
        /// </summary>

File: SimulationAgent/DeviceTelemetry/DeviceTelemetryActor.cs [hunk 4]
```diff
@@ -87,6 +91,10 @@ public enum ActorEvents
         /// </summary>
         public IDeviceClient Client => this.deviceConnectionActor.Client;
 
+        public int TotalMessagesCount => totalMessageCount;
```
Comment (avanikp) [inline SimulationAgent/DeviceTelemetry/DeviceTelemetryActor.cs]: please add summary block

        /// <summary>
       /// 
       /// </summary>

File: SimulationAgent/DeviceTelemetry/DeviceTelemetryActor.cs [hunk 5]
```diff
@@ -87,6 +91,10 @@ public enum ActorEvents
...
+
+        public int FailedMessagesCount => failedMessageCount;
```
Comment (avanikp) [inline SimulationAgent/DeviceTelemetry/DeviceTelemetryActor.cs]: same, summary block

File: SimulationAgent/SimulationRunner.cs [hunk 6]
```diff
@@ -76,14 +80,24 @@ public class SimulationRunner : ISimulationRunner
         // Flag signaling whether the simulation has started and is running (to avoid contentions)
         private bool running;
 
+        // Counter for telemetry messages
+        private int totalMessagesCount = 0;
+
+        // Counter for failed telemetry messages;
+        private int failedMessagesCount = 0;
+
+        // Per seconde counter for telemetry messages
```
Comment (avanikp) [inline SimulationAgent/SimulationRunner.cs]: spell check: seconds

File: WebService/v1/Controllers/StatusController.cs [hunk 7]
```diff
@@ -219,5 +247,33 @@ private string GetIoTHubMetricsUrl(bool isRunning)
                    $"/resourceGroups/{this.deploymentConfig.AzureResourceGroup}" +
                    $"/providers/Microsoft.Devices/IotHubs/{this.deploymentConfig.AzureIothubName}/Metrics";
         }
+
+        private int GetActiveDeviceCount(bool isRunning)
```
Comment (avanikp) [inline WebService/v1/Controllers/StatusController.cs]: Instead of passing isRunning to all functions, can we add isRunning as a private prop and use it in all places ?

Comment (avanikp) [inline WebService/v1/Controllers/StatusController.cs]: please add function summary block for this function and others

File: Services/Concurrency/RatedCounter.cs [hunk 8]
```diff
@@ -177,21 +177,40 @@ public long GetPause()
...
```
Comment (avanikp) [inline Services/Concurrency/RatedCounter.cs]: This method can be removed. GetMessagesPerSecond will return 0 if this.timestamps.count is 0.

File: SimulationAgent/DeviceTelemetry/DeviceTelemetryActor.cs [hunk 9]
```diff
@@ -60,6 +62,8 @@ public enum ActorEvents
         private string deviceId;
         private DeviceModel deviceModel;
         private long whenToRun;
+        private int totalMessageCount = 0;
+        private int failedMessageCount = 0;
```
Comment (dluc) [inline SimulationAgent/DeviceTelemetry/DeviceTelemetryActor.cs]: use `long`

Comment (dluc) [reply SimulationAgent/DeviceTelemetry/DeviceTelemetryActor.cs]: init values in the ctor

File: SimulationAgent/DeviceTelemetry/SendTelemetry.cs [hunk 10]
```diff
@@ -31,45 +32,48 @@ public void Setup(IDeviceTelemetryActor context, string deviceId, DeviceModel de
...
-                        this.context.HandleEvent(DeviceTelemetryActor.ActorEvents.TelemetryDelivered);
-                    });
+                        msg = msg.Replace("${" + value.Key + "}", value.Value.ToString());
+                    }
+
+                    this.log.Debug("Calling SendMessageAsync...",
+                        () => new { this.deviceId, MessageSchema = this.message.MessageSchema.Name, msg });
+
+                    var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
+                    await this.context.Client.SendMessageAsync(msg, this.message.MessageSchema);
```
Comment (dluc) [inline SimulationAgent/DeviceTelemetry/SendTelemetry.cs]: just before this line add:
```this.context.HandleEvent(DeviceTelemetryActor.ActorEvents.SendingTelemetry)```

File: SimulationAgent/DeviceTelemetry/DeviceTelemetryActor.cs [hunk 11]
```diff
@@ -151,10 +165,13 @@ public void HandleEvent(ActorEvents e)
                     this.ScheduleTelemetry();
                     break;
                 case ActorEvents.TelemetryDelivered:
```
Comment (dluc) [inline SimulationAgent/DeviceTelemetry/DeviceTelemetryActor.cs]: let's add a new event type called `ActorEvents.SendingTelemetry` and increment `this.totalMessageCount` only in there.  In `SendTelemetry.Run()` call `this.context.HandleEvent(DeviceTelemetryActor.ActorEvents.SendingTelemetry)` before sending the message

File: SimulationAgent/DeviceTelemetry/DeviceTelemetryActor.cs [hunk 12]
```diff
@@ -151,10 +165,13 @@ public void HandleEvent(ActorEvents e)
...
+                    this.totalMessageCount++;
                     this.actorLogger.TelemetryDelivered();
                     this.ScheduleTelemetry();
                     break;
                 case ActorEvents.TelemetryDeliveryFailed:
+                    this.totalMessageCount++;
```
Comment (dluc) [inline SimulationAgent/DeviceTelemetry/DeviceTelemetryActor.cs]: remove this

File: SimulationAgent/DeviceTelemetry/DeviceTelemetryActor.cs [hunk 13]
```diff
@@ -151,10 +165,13 @@ public void HandleEvent(ActorEvents e)
...
+                    this.failedMessageCount++;
```
Comment (dluc) [inline SimulationAgent/DeviceTelemetry/DeviceTelemetryActor.cs]: this is ok

File: SimulationAgent/DeviceTelemetry/SendTelemetry.cs [hunk 14]
```diff
@@ -31,45 +32,48 @@ public void Setup(IDeviceTelemetryActor context, string deviceId, DeviceModel de
...
+                    var timeSpent = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - now;
+                    this.log.Debug("Telemetry delivered", () => new { this.deviceId, timeSpent, MessageSchema = this.message.MessageSchema.Name });
+                    this.context.HandleEvent(DeviceTelemetryActor.ActorEvents.TelemetryDelivered);
+
```
Comment (dluc) [inline SimulationAgent/DeviceTelemetry/SendTelemetry.cs]: empty line

File: WebService.Test/v1/Controllers/StatusControllerTest.cs [hunk 15]
```diff
@@ -72,6 +75,131 @@ public StatusControllerTest(ITestOutputHelper log)
...
+
+            // Act
+            var result = await this.target.Get();
+
+            // Assert
+            Assert.Equal(MESSAGE_THROUGHPUT.ToString("F"), result.Properties["MessagesPerSecond"]);
+        }
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public async Task ItReturnsTheNumberOfTotalMessagesCount()
```
Comment (dluc) [inline WebService.Test/v1/Controllers/StatusControllerTest.cs]: can rename to: `ItReturnsTheNumberOfTotalMessages`

File: WebService.Test/v1/Controllers/StatusControllerTest.cs [hunk 16]
```diff
@@ -72,6 +75,131 @@ public StatusControllerTest(ITestOutputHelper log)
...
+            this.SetupPreprovisionedIoTHub();
+
+            // Act
+            var result = await this.target.Get();
+
+            // Assert
+            Assert.Contains("https://portal.azure.com/", result.Properties["PreprovisionedIoTHubMetricsUrl"]);
+        }
+
+
```
Comment (dluc) [inline WebService.Test/v1/Controllers/StatusControllerTest.cs]: spaces

File: Services/Concurrency/RateLimiting.cs [hunk 17]
```diff
@@ -82,5 +83,10 @@ public long GetPauseForNextMessage()
             // https://github.com/Azure/device-simulation-dotnet/issues/80
             return this.messaging.GetPause();
         }
+
+        /// <summary>
+        /// Get message throughput (messages per second)
+        /// </summary>
+        public double GetThroughputForMessages() => this.messaging.GetThroughputForMessages();
```
Comment (dluc) [inline Services/Concurrency/RateLimiting.cs]: add test

File: Services/Concurrency/RatedCounter.cs [hunk 18]
```diff
@@ -177,19 +177,37 @@ public long GetPause()
...
             if (this.log.LogLevel <= LogLevel.Debug && this.timestamps.Count > 1)
             {
-                double speed;
-                lock (this.timestamps)
+                double speed = this.GetMessagesPerSecond(this.timestamps);
+                this.log.Info(this.name, () => new { speed });
+            }
+        }
+
+        private double GetMessagesPerSecond(Queue<long> timestamps)
```
Comment (dluc) [inline Services/Concurrency/RatedCounter.cs]: remove this method, move all the logic to `GetThroughputForMessages`, and in `LogThroughput` call `GetThroughputForMessages()`

Comment (dluc) [reply Services/Concurrency/RatedCounter.cs]: for instance:
```
        /// <summary>
        /// Get messages throughput.
        /// </summary>
        /// <returns>Throughput: messages per second</returns>
        public double GetThroughputForMessages()
        {
            double speed = 0;
            lock (this.timestamps)
            {
                if (this.timestamps.Count > 1)
                {
                    // Time range in milliseconds
                    long time = this.timestamps.Last() - this.timestamps.First();

                    // Unit for speed is messages per second
                    speed = (1000 * (double) this.timestamps.Count / time * 10) / 10;
                }
            }

            return speed;
        }

        private void LogThroughput()
        {
            if (this.log.LogLevel <= LogLevel.Debug)
            {
                double speed = this.GetThroughputForMessages();
                this.log.Info(this.name, () => new { speed });
            }
        }
```

File: Services/DeviceClient.cs [hunk 19]
```diff
@@ -155,6 +155,8 @@ public Task UpdateTwinAsync(Device device)
                         Exception = e.GetType().FullName,
                         e.InnerException
                     });
+
+                throw new TelemetrySendException("Message delivery failed with " + e.Message);
```
Comment (dluc) [inline Services/DeviceClient.cs]: todo: create a github issue to add more granular checks here, i.e. we need to handle different errors in different ways.

Comment (dluc) [reply Services/DeviceClient.cs]: also, pass the original exception back
```
throw new TelemetrySendException("Message delivery failed with " + e.Message, e);
```

File: SimulationAgent/DeviceConnection/DeviceConnectionActor.cs [hunk 20]
```diff
@@ -97,6 +104,24 @@ public enum ActorEvents
...
+        /// </summary>
+        public long FailedTwinUpdatesCount => this.failedTwinUpdatesCount;
+
+        /// <summary>
+        /// Simulation error counter in DeviceConnectionActor
+        /// </summary>
+        public long SimulationErrorsCount => this.failedRegistrationsCount +
+            this.failedFetchCount +
+            this.failedTwinUpdatesCount +
+            this.failedDeviceConnectionsCount;
```
Comment (dluc) [inline SimulationAgent/DeviceConnection/DeviceConnectionActor.cs]: for all these accesses to the private fields, use the public getter if available, e.g.
```
        public long SimulationErrorsCount => this.failedRegistrationsCount +
            this.failedFetchCount +
            this.FailedTwinUpdatesCount +
            this.FailedDeviceConnectionsCount;
```

File: SimulationAgent/DeviceState/DeviceStateActor.cs [hunk 21]
```diff
@@ -56,6 +57,7 @@ public bool IsDeviceActive
         private long whenCanIUpdate;
         private int startDelayMsecs;
         private ActorStatus status;
+        private long simulationErrorsCount;
```
Comment (dluc) [inline SimulationAgent/DeviceState/DeviceStateActor.cs]: pls move all the private fields at the beginning

Comment (dluc) [reply SimulationAgent/DeviceState/DeviceStateActor.cs]: for the time being try to use this rule, we might need to refine it in future:
```
public Constant Fields
public Enums
private Constant Fields
private Fields
public Properties
public Constructors
public Methods
private Enums
private Methods
private Classes
```

File: SimulationAgent/DeviceTelemetry/SendTelemetry.cs [hunk 22]
```diff
@@ -31,45 +32,48 @@ public void Setup(IDeviceTelemetryActor context, string deviceId, DeviceModel de
             this.message = context.Message;
         }
 
-        public void Run()
+        public async void Run()
```
Comment (dluc) [inline SimulationAgent/DeviceTelemetry/SendTelemetry.cs]: let's revert this change and keep Run() a synchronous/blocking function.

File: SimulationAgent/SimulationRunner.cs [hunk 23]
```diff
@@ -411,5 +437,52 @@ private void TryToStopStateThread()
...
+        {
+            this.devicesTelemetryThread = new Thread(this.SendTelemetryThread);
+            try
+            {
+                this.devicesTelemetryThread.Start();
+            }
+            catch (Exception e)
+            {
+                this.IncreamentSimulationErrorsCount();
+                this.log.Warn("Unable to start the telemetry thread", () => new { e });
```
Comment (dluc) [inline SimulationAgent/SimulationRunner.cs]: log as an Error

File: SimulationAgent/SimulationRunner.cs [hunk 24]
```diff
@@ -411,5 +437,52 @@ private void TryToStopStateThread()
...
+        {
+            this.devicesConnectionThread = new Thread(this.ConnectDevicesThread);
+            try
+            {
+                this.devicesConnectionThread.Start();
+            }
+            catch (Exception e)
+            {
+                this.IncreamentSimulationErrorsCount();
+                this.log.Warn("Unable to start the device connection thread", () => new { e });
```
Comment (dluc) [inline SimulationAgent/SimulationRunner.cs]: log as an Error

File: SimulationAgent/SimulationRunner.cs [hunk 25]
```diff
@@ -411,5 +437,52 @@ private void TryToStopStateThread()
...
+        {
+            this.devicesStateThread = new Thread(this.UpdateDevicesStateThread);
+            try
+            {
+                this.devicesStateThread.Start();
+            }
+            catch (Exception e)
+            {
+                this.IncreamentSimulationErrorsCount();
+                this.log.Warn("Unable to start the device state thread", () => new { e });
```
Comment (dluc) [inline SimulationAgent/SimulationRunner.cs]: log as an Error

File: SimulationAgent/SimulationRunner.cs (hunk 23 above)
Comment (dluc) [reply SimulationAgent/SimulationRunner.cs]: rethrow the exception, it will be caught in `Simulation.RunAsync()`

File: SimulationAgent/SimulationRunner.cs (hunk 24 above)
Comment (dluc) [reply SimulationAgent/SimulationRunner.cs]: rethrow the exception, it will be caught in `Simulation.RunAsync()`

File: SimulationAgent/SimulationRunner.cs (hunk 25 above)
Comment (dluc) [reply SimulationAgent/SimulationRunner.cs]: rethrow the exception, it will be caught in `Simulation.RunAsync()`

File: SimulationAgent.Test/DeviceTelemetry/DeviceTelemetryActorTest.cs [hunk 26]
```diff
@@ -61,29 +61,29 @@ public void TheNumberOfFailedMessagesCountIsZeroAtStart()
...
-                this.target.HandleEvent(messageFailed);
-            }
-
-            // Get results
-            long failedMessagesCount = this.target.FailedMessagesCount;
-
-            // Assert
-            Assert.Equal(FAILED_MESSAGES_COUNT, failedMessagesCount);
-        }
+//        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
```
Comment (dluc) [inline SimulationAgent.Test/DeviceTelemetry/DeviceTelemetryActorTest.cs]: can we uncomment this test?

Comment (dluc) [reply SimulationAgent.Test/DeviceTelemetry/DeviceTelemetryActorTest.cs]: if not, please mark is as "skipped" with a justification
